src/InputOutput/writers.jl

```diff
@@ -511,6 +531,118 @@ function write!(
     )
 end
 
+"""
+    write!(
+        writer::HDF5Writer,
+        values::DataLayouts.AbstractData,
+        name::AbstractString,
+        topology::Topologies.AbstractTopology,
+    )
+
+Write an object of type 'AbstractData' and name 'name' to the HDF5 file.
+
+The `values` should belong to a `Field` whose `space`'s topology is
+`topology(axes(field))`.
+"""
+function write!(
+    writer::HDF5Writer,
+    group,
+    values::DataLayouts.AbstractData,
+    name::AbstractString,
+    topology::Topologies.AbstractTopology,
+)
+    if topology isa Topologies.Topology2D &&
+       !(writer.context isa ClimaComms.SingletonCommsContext)
+        nelems = Topologies.nelems(topology)
+        (; local_elem_gidx) = topology
+        _write_mpi!(group, values, name; nelems, local_elem_gidx)
+    else
+        _write!(group, values, name)
+    end
+end
+
+function write_plain_array!(group, array::AbstractArray, name::AbstractString)
+    nd = ndims(array)
+    dims = size(array)
+    localidx = ntuple(d -> (:), nd)
+    dataset =
+        create_dataset(group, name, datatype(eltype(array)), dataspace(dims))
+    dataset[localidx...] = array
+    return dataset
+end
+
+"""
+    _write_mpi!(
+        writer::HDF5Writer,
+        data::DataLayouts.AbstractData,
+        name::AbstractString,
+        nelems,
+        local_elem_gidx
+    )
+
+This is an internal method, meant to be used for writing data layouts to the
+HDF5 file.
+
+This method should be used when
+
+```julia
+!(topology isa Topologies.Topology2D && !(writer.context isa ClimaComms.SingletonCommsContext))
```

Comment:
Is this restriction on Topology 2D needed because we don't support MPI for Topology1D?

Reply:
Also, I think this conditional is incorrect for this case and should just be 
```
topology isa Topologies.Topology2D && !(writer.context isa ClimaComms.SingletonCommsContext)
```

Reply:
> Is this restriction on Topology 2D needed because we don't support MPI for Topology1D?

I think that is true, but I used this condition mostly to preserve the existing behavior

Reply:
It's been a while, but I think these comments came after the comments above, where you said [Okay, I looked at the existing code and think I have answered my own question about the need for two functions](https://github.com/CliMA/ClimaCore.jl/pull/2212#discussion_r1989523629), so I'm going to close. Please reopen if this is not the case.